Fmtstr no dollar payload

murph12F opened the pull request:
# No $ notation support for Fmtstr module 
Added a new feature in the fmtstr module, the fmtstr_payload(...) function now supports generating payload without using the $ notation.
This can be accomplished with the new flag in the fmstr_payload function parameters ``no_dollars``. 

> Every line of code added has effect only if the ``no_dollar`` parameter is set, if is not set the module remain unchanged. 

___
##  Function signature  
### Before
```py
def fmtstr_payload(offset, writes, numbwritten=0, write_size='byte', write_size_max='long', overflows=16, strategy="small", badbytes=frozenset(), offset_bytes=0 ) -> str  
```
### After
```py
def fmtstr_payload(offset, writes, numbwritten=0, write_size='byte', write_size_max='long', overflows=16, strategy="small", badbytes=frozenset(), offset_bytes=0, no_dollars=False)  -> str  
```

## Example: 
```py
[ins] In [5]: from pwn import *

[ins] In [6]: context.clear( arch='amd64')

[ins] In [7]: fmtstr_payload(7,{ 0xdeadbeef : 0xc00ffe }  , no_dollars=True)
Out[7]: b'%c%c%c%c%c%c%c%c%c%c%c%c%4082c%lln%194c%hhnaaaab\x00\x00\x00\x00\x00\x00\x00\x00\xef\xbe\xad\xde\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf1\xbe\xad\xde\x00\x00\x00\x00'
```

murph12F committed:
added feature to the fmtstr module, now able to generate payload without the dollar syntax.

murph12F committed:
made some improvements in the no_dollar feature implemented

murph12F committed:
reduced different lines

murph12F committed:
removed some blank lines and debug stuff

murph12F committed:
(fix) missing check for no_dollar option

murph12F committed:
(fix) updatet argument type for documentation

Arusekk commented on pwnlib/fmtstr.py:
```diff
@@ -809,7 +858,14 @@ def fmtstr_payload(offset, writes, numbwritten=0, write_size='byte', write_size_
     for _ in range(1000000):
         data_offset = (offset_bytes + len(fmt)) // context.bytes
-        fmt, data = make_payload_dollar(offset + data_offset, all_atoms, numbwritten=numbwritten)
+
+        # if no dollar is set, call the payload with the parameter no dollar set
+        if no_dollars:
+            fmt, data = make_payload_dollar(offset + data_offset, all_atoms, numbwritten=numbwritten, no_dollars=True)
+
+        else:
+            fmt, data = make_payload_dollar(offset + data_offset, all_atoms, numbwritten=numbwritten)
+
```
Why the if-else?
```suggestion
        fmt, data = make_payload_dollar(offset + data_offset, all_atoms, numbwritten=numbwritten, no_dollars=no_dollars)
```

murph12F replied:
dangggg, good one, i ll change it.

Arusekk commented on pwnlib/fmtstr.py:
```diff
@@ -709,7 +717,47 @@ def make_payload_dollar(data_offset, atoms, numbwritten=0, countersize=4):
-        fmt += "%" + str(data_offset + idx) + "$" + SPECIFIER[atom.size]
+
+            if no_dollars:
+                data += pack(0)
```
Because nullbytes often are string terminators, I would prefer some other value, like
```suggestion
                data += b'c' * context.bytes
```

Arusekk replied:
Also, it might be beneficial to use `b'c' * padding` on low values of padding (if padding < 4 on no_dollars=False and if padding < 4 + context.bytes on no_dollars=True).

murph12F replied:
i see,  fo sure, i ve just put the first one that crossed my mind, i guess that ``context.bytes`` is the lenght of address based on the arch right?

murph12F replied:
> Also, it might be beneficial to use `b'c' * padding` on low values of padding (if padding < 4 on no_dollars=False and if padding < 4 + context.bytes on no_dollars=True).

ahhh, i am not getting this part, what this about?

Arusekk replied:
> i see, fo sure, i ve just put the first one that crossed my mind, i guess that `context.bytes` is the lenght of address based on the arch right?

right, for example 8 on 64-bit and 4 on 32-bit

Arusekk replied:
> > Also, it might be beneficial to use `b'c' * padding` on low values of padding (if padding < 4 on no_dollars=False and if padding < 4 + context.bytes on no_dollars=True).
> 
> ahhh, i am not getting this part, what this about?

You can add `c` instead of `%1c`, `cc` instead of `%2c`, `ccc` instead of `%3c` and so on, if it is not longer than `%SOMETHINGc` to save off a few bytes.

murph12F replied:
cool!  what about the other thing you told me bout, didn t really got it.

murph12F replied:
sorry, tought the comments would self refresh, i ve wrote the last one w/o reading yours, i ll read now

murph12F replied:
> > > Also, it might be beneficial to use `b'c' * padding` on low values of padding (if padding < 4 on no_dollars=False and if padding < 4 + context.bytes on no_dollars=True).
> > 
> > 
> > ahhh, i am not getting this part, what this about?
> 
> You can add `c` instead of `%1c`, `cc` instead of `%2c`, `ccc` instead of `%3c` and so on, if it is not longer than `%SOMETHINGc` to save off a few bytes.

make sense now, it is convenient just if the pad is less than 3,  if the pad is 3 it takes the same amount of chars
``ccc`` - ``%3c``, right? ill add a lil fix, check it back in a second.

murph12F replied:
made it, have a look,  thx  :)

murph12F committed:
updated CHANGELOG.md

murph12F committed:
(fix) typo in the CHANGELOG.md version

Arusekk commented on pwnlib/fmtstr.py:
```diff
@@ -697,6 +698,13 @@ def make_payload_dollar(data_offset, atoms, numbwritten=0, countersize=4):
     counter = numbwritten
+
+    if no_dollars:
+        # since we cant dynamically offset, we have to increment manually the parameter index, use %c, so the numeber of bytes written is predictable
+        fmt += "%c" * ( data_offset - 1)
```
A couple typos / formatting issues

murph12F replied:
do i have to fix something about it?

Arusekk replied:
numeber, cant≠can't, space after (

murph12F replied:
are you talking bout the space missing at the end of the  parenthesis?

murph12F replied:
ok corrected them, thx , i ll close this one.

Arusekk reviewed (changes requested):
Great stuff!  Thanks a lot, would you mind adding some extra doctest that would ensure the fmtstr stuff works?  It might be beneficial to also launch the examples/fmtstr.py and examples/fmtstr2.py with no_dollars=True in the CI pipeline.

murph12F committed:
Update pwnlib/fmtstr.py

suggested change, dont need comparison in calling the make_payload_dollar

Co-authored-by: Arusekk <arek_koz@o2.pl>

murph12F committed:
Update pwnlib/fmtstr.py

suggestion; better use a non null byte thing to fill the values used by the %c to pad before writing.

Co-authored-by: Arusekk <arek_koz@o2.pl>

murph12F committed:
(fix) minor fixes, typos on comments

Arusekk commented on pwnlib/fmtstr.py:
```diff
@@ -708,8 +716,52 @@ def make_payload_dollar(data_offset, atoms, numbwritten=0, countersize=4):
-            fmt += "%" + str(padding) + "c"
-        fmt += "%" + str(data_offset + idx) + "$" + SPECIFIER[atom.size]
+            # if the padding is less than 3, it is more convenient to write it : [ len("cc") < len("%2c") ] , this could help save some bytes, if it is 3 it will take the same amout of bytes
+            if padding < 4:
+                    fmt += "c" * padding
```
I would do it this way, since on no_dollars=True we also add another pointer to the payload.
```suggestion
            if padding < 4 + context.bytes * no_dollars:
                fmt += "c" * padding
```

Arusekk commented:
How about some tests now?

murph12F commented:
> How about some tests now?

did you see the padding thing? i see that you answered that from my email but cand find your response, could you open another suggestion about that? 
sorry :/

Arusekk commented on pwnlib/fmtstr.py:
```diff
@@ -708,8 +716,52 @@ def make_payload_dollar(data_offset, atoms, numbwritten=0, countersize=4):
-            fmt += "%" + str(padding) + "c"
-        fmt += "%" + str(data_offset + idx) + "$" + SPECIFIER[atom.size]
+            # if the padding is less than 3, it is more convenient to write it : [ len("cc") < len("%2c") ] , this could help save some bytes, if it is 3 it will take the same amout of bytes
+            if padding < 4:
+                    fmt += "c" * padding
```
I would do it this way, since on no_dollars=True we also add another pointer to the payload.

```suggestion
            if padding < 4 + context.bytes * no_dollars:
                fmt += "c" * padding
```

And actually this won't work for no_dollars, you need to move it up, like 
```py
if padding < 4 + context.bytes * no_dollars:
    fmt += "c" * padding
else:
    fmt += "%" + str(padding) + "c"
    if no_dollars:
        ...
```
instead of the `if padding:` thing.

murph12F replied:
what do you mean by ``on no_dollars=True we also add another pointer to the payload.``
i am not getting why it would be convenient to 
```py
if padding < 4 + context.bytes * no_dollars:
....
```
like let s assume no_dollars is set
it would be 
```py
if padding < 12 : 
.....
```
but isn t more convient to have ``%12c``, than ``"c" * 12 `` ?

murph12F replied:
ohh i got it, if i use "c" * 2 instead of the "%2c" i ll have to not pack the "c" * 8, still not getting the 
```if padding < 4 + context.bytes * no_dollars:```

Arusekk replied:
What I meant is that it is more convenient to have `%nccccccccc%n...ptr1ptr2` than `%n%8c%n...ptr1ccccccccptr2`.

murph12F replied:
dang, you got me on this one, ill change that, gimme a sec

murph12F replied:
ok i ve done that, have a look when u can.

murph12F replied:
i ve made a quiche test seems working. but have a check just to be sure.

murph12F commented:
> > How about some tests now?
> 
> did you see the padding thing? i see that you answered that from my email but cand find your response, could you open another suggestion about that? sorry :/

i was looking at the output of the checks that failed, have a look at this, this one fail cause of the new feature added, because he expects to find %1c but we just printng "c" * 1. 
```
Document: fmtstr
----------------
**********************************************************************
File "fmtstr.rst", line ?, in default
Failed example:
    fmtstr_payload(1, {0x0: 0x00000001}, write_size='byte')
Expected:
    b'%1c%3$na\x00\x00\x00\x00'
Got:
    'c%3$naaa\x00\x00\x00\x00'
**********************************************************************
```

Arusekk commented:
Yes, the test needs to be updated.

Please add another test too, with no_dollars set to True.

murph12F committed:
update pwnlib/fmtstr.py : suggestion from Arusekk, made improvements to save some bytes in the generation of the payload

Arusekk commented:
Time for some tests :)

murph12F commented:
not really practical with tests :/ ,  what am i supposed to do, is there any doc to read about it?i ll check them next mornin,  btw, i ll go to sleep now, been awake for too many hours,  appreciate yo time mate.  I ll link with you tomorrow. good morning/night gang

murph12F commented:
just had a look before closing the laptop, are those the test that i am supposed to do?
```
   16     Examples:
   15         >>> context.clear(arch = 'amd64')
   14         >>> fmtstr_payload(1, {0x0: 0x1337babe}, write_size='int')
   13         b'%322419390c%4$llnaaaabaa\x00\x00\x00\x00\x00\x00\x00\x00'
   12         >>> fmtstr_payload(1, {0x0: 0x1337babe}, write_size='short')
   11         b'%47806c%5$lln%22649c%6$hnaaaabaa\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00'
   10         >>> fmtstr_payload(1, {0x0: 0x1337babe}, write_size='byte')
    9         b'%190c%7$lln%85c%8$hhn%36c%9$hhn%131c%10$hhnaaaab\x00\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\x
      02\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00'
    8         >>> context.clear(arch = 'i386')
    7         >>> fmtstr_payload(1, {0x0: 0x1337babe}, write_size='int')
    6         b'%322419390c%5$na\x00\x00\x00\x00'
    5         >>> fmtstr_payload(1, {0x0: 0x1337babe}, write_size='short')
    4         b'%4919c%7$hn%42887c%8$hna\x02\x00\x00\x00\x00\x00\x00\x00'
    3         >>> fmtstr_payload(1, {0x0: 0x1337babe}, write_size='byte')
    2         b'%19c%12$hhn%36c%13$hhn%131c%14$hhn%4c%15$hhn\x03\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00'
    1         >>> fmtstr_payload(1, {0x0: 0x00000001}, write_size='byte')
  855         b'%1c%3$na\x00\x00\x00\x00'
    1         >>> fmtstr_payload(1, {0x0: b"\xff\xff\x04\x11\x00\x00\x00\x00"}, write_size='short')
    2         b
    ```
sorry for lines number, vim shit

Arusekk commented:
Sure, the tests are just the code snippets in the documentation strings of the functions.  You can save time and only run relevant tests with:
```sh
$ pip install -r docs/requirements.txt
$ python -bb -m sphinx -b doctest docs/source docs/build/doctest docs/source/fmtstr.rst
```

murph12F commented:
ok, appreciate it,  i ll do that soon as i wake up.

murph12F committed:
added tests to fmtstr_payload with no dollar flag

murph12F commented:
@Arusekk added some tests ( just 3 for now ), i ve runned them as you suggested, seems working correctly. let me know, i ll keep doing some test.

murph12F commented:
Hello, i ve seen you approved the changes, and added this request to a milestone, thanks.
Could i ask why this test keep failing? : 
``coverage/coveralls — Coverage decreased (-5.3%) to 68.181% ``

also you added this to the milestone to 4.11, want me to change that in my CHANGELOG? i ve putted it in the 4.12

Arusekk commented:
Sure, also make sure the markdown is rendered correctly (i.e. one line added to the list and one to the link section).  Never mind the coverage indicator, it is broken and I have been looking for some ways to replace it.

murph12F committed:
update CHANGELOG.md with correct version

murph12F commented:
ok updated the changelog file with correct version. i ve copied the other lines to be sure to do that right, have a look in when u have a second, after this changes gets approved i ll just have to wait right, till the drop of the next milestone? ( sorry newbie in contributing other projects) appreciate it
